### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive when assertion prevents generic widening

### ISSUE THREAD

**Issue by benasher44** [typescript-eslint/typescript-eslint#12376]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.ts&code=C4TwDgpgBAKg9gO2gXigIjHAzgS2DgNwjSgB90kBzAQ3yJPLSQFdgAnagGzQG4AoPgBMIAY07U20AGbMEI-Iig4EUiGwA8MAHwAKAl2YQAXLACUJmP2FiJ0EYizAo9hIJMAjOHE4RqCfnwujlBw7gBWUKjKqmw6AN58UFDAiMZQOi6CUAD86Ji4dMRQJkwQNIVoplDUWLCpADR8AL6mAUFOwBapkSHhAHQpSPxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2R4BDTAHN0LSpHQkATAAZp0yDxpR4sACbsMrSEQ3oi6AB4tlfNZowAlKbFTsCxHQDl9z46d609AUXjJxNkxOM1oSAGtkRAAZZAAjAGEAC3RocNYOdF4AXxBsoA&tokens=false

### Repro Code

```TypeScript
type Tone = "positive" | "negative" | "neutral";

declare function infer<T>(value: T): T;
declare const cond: boolean;

const obj = infer({
  tone: (cond ? "positive" : "negative") as Tone,
});

const t: Tone = obj.tone;
```

### ESLint Config

```javascript
import tseslint from "typescript-eslint";

export default tseslint.config({
  files: ["**/*.ts"],
  languageOptions: {
    parserOptions: {
      projectService: true,
      tsconfigRootDir: import.meta.dirname,
    },
  },
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error",
  },
});
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "target": "ES2022",
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "noEmit": true,
    "skipLibCheck": true
  },
  "include": ["*.ts"]
}
```

### Expected Result

No report. The `as Tone` assertion is load-bearing for generic inference — removing it widens the conditional expression's type from `"positive" | "negative"` to `string` when TypeScript infers `T` for `infer`, so `T` becomes `{ tone: string }` instead of `{ tone: Tone }`.

### Actual Result

error  This assertion is unnecessary since the receiver accepts the original type of the expression  @typescript-eslint/no-unnecessary-type-assertion

### Additional Info

The rule checks whether the *asserted expression's* type is assignable to the receiver, but doesn't account for the fact that without the assertion, generic type inference will widen literal types before constructing the inferred parameter type.

**Comment by bradzacher** [comment]
There's nothing we can do about this.
TypeScript "backfills" the type of the call expression using the inferred type of the argument.

(1) the way TS works is that it "backfills" the type of the generic parameter using the inferred type. In a nutshell the types look exactly the same as if you had manually annotated the generic.

(2) also there's no speculative querying of types - we can only see the types for the code as written - that is it.

---

So what happens is that when we look at the assertion we ask TS for the "contextual type" of the thing that's receiving the result of the assertion `as T`. Because of (1), TS tells us the contextual type is `T` - so the rule wants to report. Because of (2) there's no way for us to ask TS "what would happen if the assertion was removed" - so we can't determine if it's a "load bearing" annotation - we really have to take it at face value.

**Comment by benasher44** [comment]
@bradzacher what about a flag to ignore casts on literals? I think that'd cover a lot of ground, in terms of not having to add in last of eslint ignore lines

**Comment by bradzacher** [comment]
That wouldn't fix anything here - because the issue isn't literals it's just cases where you want to widen or narrow the type for some reason.

For example you'd also get the same error if you did a similar "load bearing" assertion on the entire object:

```ts
const obj = infer({
  tone: (cond ? "positive" : "negative"),
} as { tone: Tone });
```
 [playground](https://typescript-eslint.io/play/#ts=6.0.3&fileType=.ts&code=C4TwDgpgBAKg9gO2gXigIjHAzgS2DgNwjSgB90kBzAQ3yJPLSQFdgAnagGzQG4AoPgBMIAY07U20AGbMEI-Iig4EUiGwA8MAHwAKAl2YQAXLACUJmP2FiJ0EYizAo9hIJMAjOHE4RqCfnwujlBw7gBWUKjKqmw6AN58UFDAiMZQOi6CUAD86Ji4dMRQJkwQNIVopgA0fAC%2BUNRYUHHJqRapULWmAUFOwO1IkSHhAHQpSPxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2R4BDTAHN0LSpHQkATAAZp0yDxpR4sACbsMrSEQ3oi6AB4tlfNZowAlKbFTsCxHQDl9z46d609AUXjJxNkxOM1oSAGtkRAAZZAAjAGEAC3RocNYOdF4AXxBsoA&tokens=false)

The issue is specifically the behaviour of using an assertion in a place where there is an inferred generic. The only way to avoid the error is to write code which avoids this case - eg annotate the generic type instead or to avoid inlining the value:

```ts
const obj = infer<{ tone: Tone }>({
  tone: (cond ? "positive" : "negative"),
});

const x = {
  tone: (cond ? "positive" : "negative") as Tone,
};
const y = infer(x);
```

[playground](https://typescript-eslint.io/play/#ts=6.0.3&fileType=.ts&code=C4TwDgpgBAKg9gO2gXigIjHAzgS2DgNwjSgB90kBzAQ3yJPLSQFdgAnagGzQG4AoPgBMIAY07U20AGbMEI-Iig4EUiGwA8MAHwAKAl2YQAXLACUJmP2FiJ0EYizAo9hIJMAjOHE4RqCfnwujlBw7gBWUKjKqhoA3lDAiMawSVAAvrqxfFAJSSY6LoJQAPzomLh0xFAmTBA0lWimADR8aaYBQU4AHpFQWTmJSPmFJWXYeIRVNVS0k41Q1FgpSC1p-J1QIL3Rajpd7QKBDk7AFqmoIAB0gxD8QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2R4BDTAHN0LSpHQkATAAZp0yDxpR4sACbsMrSEQ3oi6AB4tlfNZowAlKbFTsCxHQDl9z46d609AUXjJxNkxOM1oSAGtkRAAZZAAjAGEAC3RocNYOdF4AXxBsoA&tokens=false)
